Assets/IATK/Scripts/Model/RealtimeDataSource.cs

```diff
@@ -0,0 +1,384 @@
+            for (int i = 0; i < dimensionSizeLimit; i++)
+            {
+                dataArray[i] = 0;
+            }
+            return dataArray;
+        }
+
+        /// <summary>
+        /// Creates a dimension that can later have data set to it
+        /// </summary>
+        /// <param name="dimensionName">Sets the dimension name, used to identify the dimension (Must be unique).</param>
+        /// <param name="numberOfCategories">The data type of categories (unique values) in the data</param>
+        /// <param name="type">The data type of the dimension</param>
+        /// <returns>True if successfully added, false otherwise</returns>
+        public bool AddDimension(string dimensionName, float numberOfCategories, DataType type = DataType.String)
+        {
+            return AddDimension(dimensionName, 0, numberOfCategories - 1f, type);
+        }
+
+        /// <summary>
+        /// Creates a dimension that can later have data set to it
+        /// </summary>
+        /// <param name="dimensionName">Sets the dimension name, used to identify the dimension (Must be unique).</param>
+        /// <param name="minVal">The minimum value the dimension can hold</param>
+        /// <param name="maxVal">The maximum value the dimension can hold</param>
+        /// <param name="type">The data type of the dimension</param>
+        /// <returns>True if successfully added, false otherwise</returns>
+        public bool AddDimension(string dimensionName, float minVal, float maxVal, DataType type = DataType.Float)
+        {
+            // Don't add the dimension if it already exists
+            if (textualDimensionsList.ContainsKey(dimensionName)) return false;
+
+            var metaData = new DimensionData.Metadata();
+            metaData.minValue = minVal;
+            metaData.maxValue = maxVal;
+            metaData.type = type;
+
+            int index = dimensionData.Count;
+
+            textualDimensionsList.Add(dimensionName, new Dictionary<int, string>());
+            textualDimensionsListReverse.Add(dimensionName, new Dictionary<string, int>());
+
+            var dd = new DimensionData(dimensionName, index, metaData);
+            dd.setData(GetDefaultArray(), textualDimensionsList);
+            dimensionData.Add(dd);
+            dimensionPointers.Add(0);
+
+            //Debug.Log("RTDS AddDimension => " + dd.Identifier + ", " + dd.Index);
+            return true;
+        }
+
+        /// <summary>
+        /// Sets a data value by index
+        /// </summary>
+        /// <param name="index">Index of dimension</param>
+        /// <param name="val">Value to set the dimension</param>
+        /// <returns>True if successfully set, false otherwise</returns>
+        public bool SetData(int index, float val)
+        {
+            return SetData(this[index].Identifier, val);
+        }
+
+
+        /// <summary>
+        /// This is important otherwise the overloading can NOT be resolved from outside (eg JS)
```

Comment:
When I have time I'll take a look at the JS code to see if I can find a work around. But if I can't then I'm happy to keep these functions if you need them 🙂

Reply:
This mostly goes back to namespace exposition. In the way it is done, js does not care too much about the parameterlist of an functions. Or if the function is a real js function or a bound c# function. Nullable types might get autocasted. Depending which how functions are oerloaded it might be hard to resolve. For example, the way it implemented by the JS Interpreted, is that you are able to call a two-param function with three params. Like `foo(a, b) {...}` might be called as `foo(1,2,3)` and this could be resolved. It is not hard to tell, that following that schemes, overloading might cause some pain.

The second point is that non-static, never-called function on the c# layer might not be exposed. Of course there is a lot more to it, feel free to check out PowerUi or my fork. 

Having an old school c99 styled semi-global function, allows a clean resolution. I thought this might be interesting to you.

Reply:
Yes it is very interesting, I've done a little research into it but not sure there is any better solution than the one you found. I plan on taking better look at it in the next few days though just to check.

Also thank for give more detail on the issue, that is appreciated 🙂